File: 110-balanced-binary-tree/110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root) {
        
        if (!root) return 0;
        
        if (!root->left && !root->right) return 1;
        
        int left = solve(root->left);
        int right = solve(root->right);
        
        if (left == -1 || right == -1) return -1;
        
        if (abs(left - right)  > 1) return -1;
        
        return max(solve(root->left), solve(root->right)) + 1;
    }
    
    bool isBalanced(TreeNode* root) {
        
        
        
        return solve(root) != -1;
    }
};

File: 110-balanced-binary-tree/110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        return self.solve(root) != -1
 
    def solve(self, node: Optional[TreeNode]) -> int:
        
        if not node:
            return 0
        
        if not (node.left or node.right):
            return 1
        
        lv, rv = self.solve(node.left), self.solve(node.right)
        
        if lv < 0 or rv < 0 or abs(lv - rv) > 1:
            return -1
        
        return max(lv, rv) + 1